assembly/src/errors.rs

```diff
@@ -71,6 +71,14 @@ pub enum AssemblyError {
         #[source_code]
         source_file: Option<Arc<SourceFile>>,
     },
+
+    #[error("invalid procedure: body must contain at least one instruction if it has decorators")]
+    #[diagnostic()]
+    EmptyProcedureBodyWithDecorators {
+        span: SourceSpan,
```

Comment:
This is an example of an error where using `Report` and ad-hoc diagnostics works better than a concrete error type like `AssemblyError`. This error sets the stage for what is wrong, but is sort of incomplete from the user perspective. Ideally, you'd have the primary label span the procedure body, with a secondary label spanning the first decorator, with a message like `'<opcode>' is a decorator, which cannot appear in a procedure body without at least one non-decorator instruction`, with help text attached to the diagnostic like `try adding a 'nop' instruction at the end of the procedure body to anchor the decorators`.

By using `AssemblyError` for things like this, we make it difficult (if not impossible) to construct these kind of helpful diagnostics, which is why most APIs are designed to return `Report` rather than `AssemblyError`.

Your call on whether you want to make the change, but we should try to get in the habit of thinking about how these errors will be seen and acted on by users, particularly those that don't have as strong a familiarity with Miden Assembly, and may not understand the difference between things like decorators and instructions, since they appear the same way in the source code.

Reply:
I echoed this comment in https://github.com/0xPolygonMiden/miden-vm/issues/1431 so that we can address it when removing `AssemblyError`.